programs/library-bridge/LibraryBridgeHandlers.cpp

```diff
@@ -399,4 +410,202 @@ void ExternalDictionaryLibraryBridgeExistsHandler::handleRequest(HTTPServerReque
+            {
+                processError(response, "No 'library_path' in request URL");
+                return;
+            }
+
+            const String & library_path = params.get("library_path");
+
+            if (!params.has("model_path"))
+            {
+                processError(response, "No 'model_path' in request URL");
+                return;
+            }
+
+            const String & model_path = params.get("model_path");
+
+            CatBoostLibraryHandlerFactory::instance().remove(model_path);
+
+            CatBoostLibraryHandlerFactory::instance().create(library_path, model_path);
+
+            auto catboost_handler = CatBoostLibraryHandlerFactory::instance().get(model_path);
+
+            if (!catboost_handler)
+            {
+                processError(response, "CatBoost library is not loaded for model " + model_path);
+                return;
+            }
+
+            size_t tree_count = catboost_handler->getTreeCount();
+            writeIntBinary(tree_count, out);
+        }
+        else if (method == "catboost_libEvaluate")
+        {
+            auto & read_buf = request.getStream();
+            params.read(read_buf);
+
+            if (!params.has("model_path"))
+            {
+                processError(response, "No 'model_path' in request URL");
+                return;
+            }
+
+            const String & model_path = params.get("model_path");
+
+            if (!params.has("data"))
+            {
+                processError(response, "No 'data' in request URL");
+                return;
+            }
+
+            const String & data = params.get("data");
+
+            ReadBufferFromString string_read_buf(data);
+            NativeReader deserializer(string_read_buf, 0);
+            Block block_read = deserializer.read();
+
+            Columns col_ptrs = block_read.getColumns();
+            ColumnRawPtrs col_raw_ptrs;
+            for (const auto & p : col_ptrs)
+                col_raw_ptrs.push_back(&*p);
+
+            auto catboost_handler = CatBoostLibraryHandlerFactory::instance().get(model_path);
+
+            if (!catboost_handler)
+            {
+                processError(response, "CatBoost library is not loaded for model" + model_path);
+                return;
+            }
+
+            ColumnPtr res_col = catboost_handler->evaluate(col_raw_ptrs);
+
+            DataTypePtr res_col_type = std::make_shared<DataTypeFloat64>();
+            String res_col_name = "res_col";
+
+            ColumnsWithTypeAndName res_cols_with_type_and_name;
+            res_cols_with_type_and_name.push_back({res_col, res_col_type, res_col_name});
+
+            WriteBufferFromOwnString string_write_buf;
+            Block block_write(res_cols_with_type_and_name);
+            NativeWriter native_writer{string_write_buf, 0, block_write};
+            native_writer.write(block_write);
+
+            writeStringBinary(string_write_buf.str(), out);
```

Comment:
I am actually not sure in what way `writeData` is more efficient than this code.

I suppose, since the exchanged data is typically not very big (it is ultimately limited by the maximum URL length) and since the data is cleared after each call to SQL function `catboostEvaluate`, we can do without further optimization for now.

Reply:
> I am actually not sure in what way writeData is more efficient than this code.

currently you write the block twice - first you put it into string via WriteBufferFromOwnString and then write it to `out`. But you could write block to `out` at once without writing it into string.

Reply:
Done, thanks.